src/Dictionaries/CacheDictionary.inc.h

```diff
@@ -277,69 +295,101 @@ void CacheDictionary::getItemsString(
         }
     }
 
-    ProfileEvents::increment(ProfileEvents::DictCacheKeysExpired, cache_expired_ids.size());
-    ProfileEvents::increment(ProfileEvents::DictCacheKeysNotFound, cache_not_found_ids.size());
+    ProfileEvents::increment(ProfileEvents::DictCacheKeysExpired, cache_expired_count);
+    ProfileEvents::increment(ProfileEvents::DictCacheKeysNotFound, cache_not_found_count);
     ProfileEvents::increment(ProfileEvents::DictCacheKeysHit, cache_hit);
 
     query_count.fetch_add(rows, std::memory_order_relaxed);
-    hit_count.fetch_add(rows - cache_expired_ids.size() - cache_not_found_ids.size(), std::memory_order_release);
+    hit_count.fetch_add(rows - cache_expired_count - cache_not_found_count, std::memory_order_release);
 
     /// Async update of expired keys.
-    if (cache_not_found_ids.empty())
+    if (!cache_not_found_count)
     {
-        if (allow_read_expired_keys && !cache_expired_ids.empty())
+        if (allow_read_expired_keys && cache_expired_count)
         {
             std::vector<Key> required_expired_ids;
-            required_expired_ids.reserve(cache_not_found_ids.size());
-            std::transform(std::begin(cache_expired_ids), std::end(cache_expired_ids),
+            required_expired_ids.reserve(cache_expired_count);
+            std::transform(std::begin(cache_expired_or_not_found_ids), std::end(cache_expired_or_not_found_ids),
                            std::back_inserter(required_expired_ids), [](auto & pair) { return pair.first; });
 
-            auto update_unit_ptr = std::make_shared<UpdateUnit>(required_expired_ids);
+            auto update_unit_ptr = std::make_shared<UpdateUnit>(std::move(required_expired_ids));
 
             tryPushToUpdateQueueOrThrow(update_unit_ptr);
 
-            /// Do not return at this point, because there some extra stuff to do at the end of this method.
+            /// Insert all found keys and defaults to output array.
+            out->getChars().reserve(total_length);
+
+            for (const auto row : ext::range(0, ext::size(ids)))
+            {
+                const auto id = ids[row];
+                StringRef value;
+
+                /// Previously we stored found keys in map.
+                const auto it = local_cache.find(id);
+                if (it != local_cache.end())
+                    value = StringRef(it->second);
+
+                value = get_default(row);
```

Comment:
Does this mean we always insert the default value?

Reply:
Missed else.